src/coreclr/src/System.Private.CoreLib/src/System/StubHelpers.Unix.cs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class OleAut32
+    {
+        internal static unsafe IntPtr SysAllocStringByteLen(byte[]? str, uint length)
+        {
+            // SysAllocString on Windows aligns the memory block size up
+            const nuint WIN32_ALLOC_ALIGN = 15;
+
+            ulong cbNative = (ulong)(uint)length + (uint)sizeof(IntPtr) + (uint)sizeof(char) + WIN32_ALLOC_ALIGN;
+            if (cbNative > uint.MaxValue)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            IntPtr p = Interop.Sys.MemAlloc((nuint)cbNative & ~WIN32_ALLOC_ALIGN);
+            if (p == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            IntPtr s = p + sizeof(IntPtr);
+            *(((uint*)s) - 1) = (uint)(length * sizeof(char));
+
+            if (str is not null)
+            {
+                fixed (byte* firstChar = str)
+                {
+                    string.wstrcpy((char*)p, (char*)firstChar, str.Length + 1);
+                }
+            }
+
+            // NULL-terminate with both a narrow and wide zero.
+            *((char*)p + length) = '\0';
+            *(uint*)((char*)p + ((length + 1) & ~1)) = 0;
```

Comment:
The sizes here do not match the C++ implementation. C# `char` is 16 bits but C++ `char` is 8 bits. C# `uint` is 32 bits but C++ `WCHAR` is 16 bits. <https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/palrt/bstr.cpp#L150-L152>